File: 7_graph/带权图最短路和最小生成树/Bellman_ford/743. 网络延迟时间.ts
/**
 * @param {number[][]} times
 * @param {number} n  有 n 个网络节点，标记为 1 到 n。
 * @param {number} k
 * @return {number}
 */
function networkDelayTime(times: number[][], n: number, k: number): number {
  const dist = Array<number>(n + 1).fill(Infinity)
  dist[k] = 0

  for (let i = 0; i < n; i++) {
    for (const [u, v, w] of times) {
      if (dist[u] + w < dist[v]) {
        dist[v] = dist[u] + w
      }
    }
  }

  let res = -1
  for (let i = 1; i <= n; i++) {
    res = Math.max(res, dist[i])
  }

  return res === Infinity ? -1 : res
}

export default 1

console.log(
  networkDelayTime(
    [
      [2, 1, 1],
      [2, 3, 1],
      [3, 4, 1],
    ],
    4,
    2
  )
)

File: 7_graph/带权图最短路和最小生成树/dijkstra单源/743. 网络延迟时间.py
from collections import defaultdict
from heapq import heappop, heappush
from typing import DefaultDict, List, Optional, Tuple, Union, overload

INF = int(1e20)


@overload
def dijkstra(n: int, adjMap: DefaultDict[int, DefaultDict[int, int]], start: int) -> List[int]:
    ...


@overload
def dijkstra(n: int, adjMap: DefaultDict[int, DefaultDict[int, int]], start: int, end: int) -> int:
    ...


def dijkstra(
    n: int, adjMap: DefaultDict[int, DefaultDict[int, int]], start: int, end: Optional[int] = None
) -> Union[int, List[int]]:
    dist = [INF] * n
    dist[start] = 0
    pq = [(0, start)]
    while pq:
        curDist, cur = heappop(pq)
        if dist[cur] < curDist:
            continue
        if end is not None and cur == end:
            return curDist
        for next in adjMap[cur]:
            if dist[next] > dist[cur] + adjMap[cur][next]:
                dist[next] = dist[cur] + adjMap[cur][next]
                heappush(pq, (dist[next], next))
    return dist


class Solution:
    def networkDelayTime(self, times: List[List[int]], n: int, k: int) -> int:
        adjMap = defaultdict(lambda: defaultdict(lambda: INF))
        for u, v, w in times:
            adjMap[u - 1][v - 1] = w
        dist = dijkstra(n, adjMap, k - 1)
        return res if (res := max(dist)) < INF else -1

